File: java/src/test/java/com/kcl/AstJsonAlignmentTest.java
package com.kcl.ast;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

import org.junit.Test;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.kcl.api.API;
import com.kcl.api.Spec.ParseFileArgs;
import com.kcl.api.Spec.ParseFileResult;
import com.kcl.api.Spec.ParseProgramArgs;
import com.kcl.api.Spec.ParseProgramResult;
import com.kcl.util.JsonUtil;

/**
 * Round-trip tests that verify the Java AST classes match the JSON shape
 * produced by the Rust compiler in {@code ../kcl/crates/ast/src/ast.rs}.
 *
 * <p>Each test parses a real KCL source file via {@code parseFile} /
 * {@code parseProgram} and asserts that the resulting {@code ast_json}
 * deserializes cleanly into the Java AST classes with the structure we expect.
 *
 * <p>If the Java AST drifts from the Rust schema — wrong field names, missing
 * variants, wrong discriminator values — the deserialization fails with a
 * clear message, or an assertion fails.
 */
public class AstJsonAlignmentTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final String FIXTURE =
            "src/test_data/ast_alignment/main.k";

    private static String parseAstJson(String relativePath) throws Exception {
        Path path = Paths.get(relativePath);
        API api = new API();
        ParseFileResult result = api.parseFile(
                ParseFileArgs.newBuilder().setPath(path.toString()).build());
        return result.getAstJson();
    }

    private static String parseProgramAstJson(String relativePath) throws Exception {
        Path path = Paths.get(relativePath);
        API api = new API();
        ParseProgramResult result = api.parseProgram(
                ParseProgramArgs.newBuilder().addPaths(path.toString()).build());
        return result.getAstJson();
    }

    // --- Module ------------------------------------------------------------

    @Test
    public void module_parsesAndHasNoPkgField() throws Exception {
        String json = parseAstJson(FIXTURE);
        Module m = MAPPER.readValue(json, Module.class);
        // parseFile resolves relative paths to absolute ones before handing
        // them to the parser, so `m.getFilename()` carries the absolute form.
        assertTrue("expected filename to end with the relative fixture path",
                m.getFilename().endsWith(FIXTURE));
        assertNotNull(m.getBody());

        // Rust Module has no `pkg` field — the serialized JSON must not contain
        // one, and Java's Module class must not produce one either.
        JsonNode tree = MAPPER.readTree(json);
        assertFalse("Module JSON must not contain 'pkg': " + json,
                tree.has("pkg"));

        String reserialized = MAPPER.writeValueAsString(m);
        assertFalse("Java Module must not serialize 'pkg': " + reserialized,
                reserialized.contains("\"pkg\""));
    }

    // --- NumberLit/StringLit/NameConstantLit discriminator ----------------

    @Test
    public void literalDiscriminators_useLongFormNotLiteralEnumForm() throws Exception {
        String json = parseAstJson(FIXTURE);
        JsonNode root = MAPPER.readTree(json);

        // Walk the body and confirm every literal node uses the long-form
        // discriminator ("NumberLit"/"StringLit"/"NameConstantLit") instead of
        // the inner Literal-enum form ("Number"/"String"/"NameConstant").
        for (JsonNode stmt : root.path("body")) {
            checkLiteralDiscriminators(stmt);
        }
    }

    private void checkLiteralDiscriminators(JsonNode node) {
        String type = node.path("type").asText("");
        if (type.equals("NumberLit") || type.equals("StringLit")
                || type.equals("NameConstantLit")) {
            // ok — this is the form we want
        } else if (type.equals("Number") || type.equals("String")
                || type.equals("NameConstant")) {
            throw new AssertionError(
                    "Rust emits Expr variant discriminators ('NumberLit', "
                            + "'StringLit', 'NameConstantLit'); got short form '"
                            + type + "' in " + node);
        }
        // recurse into children that may carry Expr nodes
        for (String key : new String[]{"value", "node", "expr", "if_cond",
                "test", "msg", "arg", "left", "right", "body", "orelse",
                "operands", "func", "cond"}) {
            if (node.has(key)) {
                checkLiteralDiscriminators(node.path(key));
            }
        }
        if (node.isArray()) {
            for (JsonNode child : node) {
                checkLiteralDiscriminators(child);
            }
        }
    }

    // --- ConfigEntry.is_shorthand -----------------------------------------

    @Test
    public void configEntry_supportsIsShorthand() throws Exception {
        // Round-trip a ConfigEntry literal: when the Rust compiler emits
        // `is_shorthand`, the field must be preserved; when absent, it must
        // default to false.
        String withShorthand = "{"
                + "\"key\":null,"
                + "\"value\":null,"
                + "\"operation\":\"Union\","
                + "\"is_shorthand\":true"
                + "}";
        ConfigEntry ce1 = MAPPER.readValue(withShorthand, ConfigEntry.class);
        assertTrue(ce1.isShorthand());

        String withoutShorthand = "{"
                + "\"key\":null,"
                + "\"value\":null,"
                + "\"operation\":\"Union\""
                + "}";
        ConfigEntry ce2 = MAPPER.readValue(withoutShorthand, ConfigEntry.class);
        assertFalse(ce2.isShorthand());
    }

    // --- UnificationStmt.value must be SchemaExpr (not SchemaConfig) ------

    @Test
    public void unificationStmt_valueIsSchemaExpr() throws Exception {
        // x = Person {name = "Alice", age = 30}
        // produces an AssignStmt with value being a SchemaExpr.
        String json = parseAstJson(FIXTURE);
        JsonNode root = MAPPER.readTree(json);

        JsonNode personAssign = findStmtWithName(root, "x");
        assertNotNull("expected AssignStmt to `x`", personAssign);

        JsonNode value = personAssign.path("value").path("node");
        assertEquals("Schema", value.path("type").asText());
        // Sanity check: schema's `config` field (the inline body) is non-null.
        assertTrue(value.path("config") != null);
    }

    @Test
    public void schemaStmt_decoratorsUseFlatDecoratorStruct() throws Exception {
        // The Java binding intentionally models decorators as a flat
        // `Decorator` struct (not the polymorphic CallExpr variant) so that
        // nested-in-NodeRef deserialization does not depend on a `type`
        // discriminator. Verify the wire JSON contains `func`/`args`/`keywords`
        // keys (the Decorator shape), not a `"type":"Call"` tag.
        String json = parseAstJson(FIXTURE);
        JsonNode root = MAPPER.readTree(json);

        JsonNode articleSchema = findSchemaByName(root, "Article");
        assertNotNull("expected Article schema", articleSchema);

        JsonNode decorators = articleSchema.path("decorators");
        assertTrue("Article should have at least one decorator",
                decorators.isArray() && decorators.size() > 0);

        for (JsonNode decorator : decorators) {
            JsonNode node = decorator.path("node");
            // Rust emits a flat {func, args, keywords} object — no "type" tag.
            assertFalse("Decorator node must not carry a 'type' tag: " + node,
                    node.has("type"));
            assertTrue("Decorator must carry 'func': " + node, node.has("func"));
        }
    }

    @Test
    public void schemaStmt_deserializesIntoJavaClass() throws Exception {
        // Verify the JSON deserializes into SchemaStmt without errors and
        // exposes decorators as a typed List<NodeRef<Decorator>>.
        String json = parseAstJson(FIXTURE);
        JsonNode root = MAPPER.readTree(json);
        JsonNode articleSchemaNode = findSchemaByName(root, "Article");
        assertNotNull(articleSchemaNode);

        Module m = MAPPER.readValue(json, Module.class);
        SchemaStmt article = (SchemaStmt) m.getBody().stream()
                .map(n -> n.getNode())
                .filter(n -> n instanceof SchemaStmt
                        && ((SchemaStmt) n).getName() != null
                        && "Article".equals(((SchemaStmt) n).getName().getNode()))
                .findFirst().orElseThrow(
                        () -> new AssertionError("Article schema not found"));
        assertNotNull(article);
        List<NodeRef<Decorator>> decorators = article.getDecorators();
        assertNotNull(decorators);
        assertTrue(decorators.size() > 0);
        for (NodeRef<Decorator> d : decorators) {
            assertNotNull(d.getNode());
            assertTrue(d.getNode() instanceof Decorator);
        }
    }

    @Test
    public void schemaAttr_hasDecoratorsField() throws Exception {
        // Verify SchemaAttr supports the `decorators` field, since Rust
        // declares it as `Vec<NodeRef<CallExpr>>`.
        String json = parseAstJson(FIXTURE);
        Module m = MAPPER.readValue(json, Module.class);

        // Iterate body and find Person schema, then its `name` attribute.
        for (NodeRef<Stmt> wrapped : m.getBody()) {
            Stmt stmt = wrapped.getNode();
            if (stmt instanceof SchemaStmt) {
                SchemaStmt schema = (SchemaStmt) stmt;
                if (schema.getName() != null
                        && "Person".equals(schema.getName().getNode())) {
                    boolean foundName = false;
                    for (NodeRef<Stmt> attrWrapped : schema.getBody()) {
                        Stmt attr = attrWrapped.getNode();
                        if (attr instanceof SchemaAttr
                                && ((SchemaAttr) attr).getName() != null
                                && "name".equals(
                                        ((SchemaAttr) attr).getName()
                                                .getNode())) {
                            SchemaAttr nameAttr = (SchemaAttr) attr;
                            assertNotNull(nameAttr.getDecorators());
                            // @deprecated has no args.
                            assertEquals(1, nameAttr.getDecorators().size());
                            assertNotNull(
                                    nameAttr.getDecorators().get(0).getNode());
                            foundName = true;
                        }
                    }
                    assertTrue("expected to find `name` SchemaAttr", foundName);
                    return;
                }
            }
        }
        throw new AssertionError("Person schema not found");
    }

    // --- Program-level parse ---------------------------------------------

    @Test
    public void parseProgram_returnsValidModulePerFile() throws Exception {
        String json = parseProgramAstJson(FIXTURE);
        com.kcl.ast.Program program = JsonUtil.deserializeProgram(json);
        assertNotNull(program);
        assertNotNull(program.getMainPackage());
        assertTrue(program.getMainPackage().size() > 0);
    }

    // --- helpers ----------------------------------------------------------

    private static JsonNode findStmtWithName(JsonNode moduleNode, String name) {
        for (JsonNode stmt : moduleNode.path("body")) {
            // Look for AssignStmt with the matching target name.
            JsonNode node = stmt.path("node");
            if ("Assign".equals(node.path("type").asText())) {
                JsonNode targets = node.path("targets");
                if (targets.isArray() && targets.size() > 0) {
                    JsonNode targetName = targets.path(0).path("node")
                            .path("name").path("node");
                    if (name.equals(targetName.asText())) {
                        return node;
                    }
                }
            }
        }
        return null;
    }

    private static JsonNode findSchemaByName(JsonNode moduleNode, String name) {
        for (JsonNode stmt : moduleNode.path("body")) {
            JsonNode node = stmt.path("node");
            if ("Schema".equals(node.path("type").asText())) {
                JsonNode schemaName = node.path("name").path("node");
                if (name.equals(schemaName.asText())) {
                    return node;
                }
            }
        }
        return null;
    }
}

File: kotlin/src/test/kotlin/com/kcl/AstJsonAlignmentTest.kt
// AstJsonAlignmentTest.kt — Round-trip AST alignment tests for the Kotlin binding.
//
// Mirrors the Java `AstJsonAlignmentTest`, Go `TestAstJsonAlignment`,
// Python `tests/ast_test.py`, Node.js `__test__/ast_alignment.spec.mjs`,
// .NET `KclLib.Tests/AstAlignmentTest.cs`, and WASM
// `tests/ast_alignment.test.ts`: parse a real KCL fixture through the
// native JNI bridge (`API.parseFile`/`API.parseProgram`) and verify the
// resulting `astJson` string deserializes cleanly into the typed AST
// classes in `com.kcl.ast`.

package com.kcl

import com.kcl.api.API
import com.kcl.ast.Decorator
import com.kcl.ast.Module
import com.kcl.ast.SchemaAttr
import com.kcl.ast.SchemaStmt
import com.kcl.ast.parseModule
import com.kcl.ast.parseProgram
import com.kcl.api.Spec.ParseFileArgs
import com.kcl.api.Spec.ParseFileResult
import com.kcl.api.Spec.ParseProgramArgs
import com.kcl.api.Spec.ParseProgramResult
import java.nio.file.Paths
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class AstJsonAlignmentTest {
    companion object {
        private val FIXTURE_PATH = Paths.get("src", "test_data", "ast_alignment", "main.k").toString()
        private val api = API()

        private fun firstSchemaByName(module: Module, name: String): SchemaStmt? =
            module.body
                .map { it.node }
                .filterIsInstance<SchemaStmt>()
                .firstOrNull { it.name?.node == name }

        private fun firstAssignNamed(module: Module, name: String): com.kcl.ast.AssignStmt? =
            module.body
                .map { it.node }
                .filterIsInstance<com.kcl.ast.AssignStmt>()
                .firstOrNull { stmt ->
                    val target = stmt.targets.firstOrNull()?.node
                    target?.name == name
                }
    }

    @Test
    fun `module has filename and no pkg`() {
        val result: ParseFileResult = api.parseFile(ParseFileArgs.newBuilder().setPath(FIXTURE_PATH).build())
        assertEquals(0, result.errorsCount)
        val module = parseModule(result.astJson)
        assertTrue(module.filename.endsWith("main.k"))
        // Module must not carry a `pkg` field — round-trip should produce
        // neither one in input nor one on output.
        assertFalse(
            module::class.java.declaredFields.any { it.name == "pkg" },
            "Module.java should not have a pkg field after alignment",
        )
    }

    @Test
    fun `literal discriminators use long form`() {
        // The fixture declares string literals (`"anonymous"`, `"Alice"`)
        // inside SchemaAttr defaults and SchemaExpr configs. The Rust
        // compiler emits them with the long-form `"StringLit"` discriminator;
        // Jackson resolves that to `com.kcl.ast.StringLit` via the
        // `@JsonSubTypes` mapping on `Expr`. A legacy short-form `"String"`
        // tag (from `@JsonTypeName("String")` on the concrete class) would
        // fail the polymorphic match.
        val result: ParseFileResult = api.parseFile(ParseFileArgs.newBuilder().setPath(FIXTURE_PATH).build())
        val astJson = result.astJson
        assertTrue(astJson.contains("\"StringLit\""), "expected \"StringLit\" tag in wire JSON")
        // And the long-form discriminator survives round-trip — deserializing
        // the AST into the typed classes must yield at least one `StringLit`
        // (e.g. the `name: str = "anonymous"` default on `Person`).
        val module = parseModule(astJson)
        val hasStringLit = walkForStringLit(module.body)
        assertNotNull(hasStringLit, "expected at least one StringLit in fixture body")
    }

    private fun walkForStringLit(
        items: List<com.kcl.ast.NodeRef<com.kcl.ast.Stmt>>,
    ): com.kcl.ast.StringLit? {
        for (item in items) {
            val node = item.node
            // SchemaAttr.value is `NodeRef<Expr>` whose `node` is the Expr
            // polymorphic instance — recurse into all Expr-bearing fields
            // by visiting each SchemaStmt's body.
            if (node is com.kcl.ast.SchemaStmt) {
                // Walk the SchemaStmt.body recursively (lambdas in check
                // expressions, nested schemas, etc.).
                val found = walkForStringLit(node.body)
                if (found != null) return found
                // And inspect each SchemaAttr's default value for a
                // StringLit directly.
                for (attrRef in node.body) {
                    val attr = attrRef.node
                    if (attr is com.kcl.ast.SchemaAttr) {
                        val lit = attr.value?.node
                        if (lit is com.kcl.ast.StringLit) return lit
                    }
                }
            }
        }
        return null
    }

    @Test
    fun `configEntry isShorthand round-trips`() {
        // Mirror Rust's #[serde(skip_serializing_if = "is_false")]: omitted
        // when false, emitted when true. The Java/Kotlin AST exposes this
        // as a boolean defaulting to false.
        val ce = com.kcl.ast.ConfigEntry()
        assertFalse(ce.isShorthand)
        ce.isShorthand = true
        assertTrue(ce.isShorthand)
    }

    @Test
    fun `assign stmt with schema expr value`() {
        val result: ParseFileResult = api.parseFile(ParseFileArgs.newBuilder().setPath(FIXTURE_PATH).build())
        val module = parseModule(result.astJson)
        val assign = firstAssignNamed(module, "x")
        assertNotNull(assign)
        val value = assign!!.value.node
        assertTrue(
            value is com.kcl.ast.SchemaExpr,
            "expected SchemaExpr on the right-hand side of `x =`, got ${value::class.simpleName}",
        )
    }

    @Test
    fun `schema stmt decorators are flat Decorator DTO`() {
        val result: ParseFileResult = api.parseFile(ParseFileArgs.newBuilder().setPath(FIXTURE_PATH).build())
        val module = parseModule(result.astJson)
        val article = firstSchemaByName(module, "Article")
        assertNotNull(article)
        val decorators: List<Decorator> = article!!.decorators.map { it.node }
        assertTrue(decorators.isNotEmpty())
        // The Decorator.Func payload is a Node wrapping an Identifier expression
        // (no `"type":"Call"` tag in the flat shape).
        decorators.forEach { deco ->
            assertTrue(
                deco.func.node is com.kcl.ast.IdentifierExpr,
                "Decorator.func should resolve to IdentifierExpr, got ${deco.func.node::class.simpleName}",
            )
        }
    }

    @Test
    fun `schema attr has decorators field`() {
        val result: ParseFileResult = api.parseFile(ParseFileArgs.newBuilder().setPath(FIXTURE_PATH).build())
        val module = parseModule(result.astJson)
        val person = firstSchemaByName(module, "Person")
        assertNotNull(person)
        val nameAttr: SchemaAttr? = person!!.body
            .map { it.node }
            .filterIsInstance<SchemaAttr>()
            .firstOrNull { it.name.node == "name" }
        assertNotNull(nameAttr)
        assertEquals(1, nameAttr!!.decorators.size)
    }

    @Test
    fun `lambda expr with arguments`() {
        val result: ParseFileResult = api.parseFile(ParseFileArgs.newBuilder().setPath(FIXTURE_PATH).build())
        val module = parseModule(result.astJson)
        val adder = firstAssignNamed(module, "adder")
        assertNotNull(adder)
        val value = adder!!.value.node
        assertTrue(value is com.kcl.ast.LambdaExpr)
        val args = (value as com.kcl.ast.LambdaExpr).args.node.args
        assertEquals(2, args.size)
    }

    @Test
    fun `parseProgram returns list of modules`() {
        val result: ParseProgramResult =
            api.parseProgram(ParseProgramArgs.newBuilder().addPaths(FIXTURE_PATH).build())
        assertEquals(0, result.errorsCount)
        val modules = parseProgram(result.astJson)
        assertTrue(modules.isNotEmpty())
        assertTrue(modules.first().filename.endsWith(".k"))
    }
}